### DISCUSSION TITLE
Unit testing unique constraint on create

### DISCUSSION THREAD

**Discussion by AnimeAllstar** [prisma/prisma#17736] [Q&A] [answered]
I was writing unit tests to for a POST request that creates a user. I wanted to test whether an error is thrown when prisma creates a second user that violates for unique constraint on email. 

However, I'm mocking my prisma client using singleton.ts as specified [here](https://www.prisma.io/docs/guides/testing/unit-testing) due to which there is no record actually being created. Is there a way for me to test the response sent on a failure to create a user due to the unique constraint being violated using a mocked client?

What I'd ideally want to test using the mock client is:
```
create user1 works
create user2 with the same email as user1 (violates unique constraint) fails
```

Will I just have to test on an actual client?

Another solution that I had was to make a call to findFirst and check if a user with that email exists, but that is two calls to prisma instead of one.

**Comment by nurul3101** [comment, marked as answer]
Hello @AnimeAllstar 👋 

For your use case, I would say not mocking PrismaClient and actually creating a record would be ideal to validate that everything works as expected. Another alternative would be to mock the failure and throw an explicit error of `Email already exists` using [MockRejectedValue](https://jestjs.io/docs/mock-function-api#mockfnmockrejectedvaluevalue)

On a side note, I would recommend you to have a look at this new Testing series that we published, this should be helpful to you: https://www.prisma.io/blog/testing-series-2-xPhjjmIEsM

**Reply by AnimeAllstar** [reply]
Thanks

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
